webp/view/src/main/cpp/webp_view.cpp

```diff
@@ -0,0 +1,280 @@
+
+#define  LOG_TAG    "libwebp-view"
+#define  LOGI(...)  __android_log_print(ANDROID_LOG_INFO,LOG_TAG,__VA_ARGS__)
+#define  LOGW(...)  __android_log_print(ANDROID_LOG_WARN,LOG_TAG,__VA_ARGS__)
+#define  LOGE(...)  __android_log_print(ANDROID_LOG_ERROR,LOG_TAG,__VA_ARGS__)
+
+/*
+ * webp files that are inside assets folder:
+ *    they will be decoded and displayed as slide-show
+ *    decoding will happen in its own thread
+ */
+const char * frames[] = {
+        "clips/frame1.webp",
+        "clips/frame2.webp",
+        "clips/frame3.webp",
+};
+const int kFRAME_COUNT = sizeof(frames) / sizeof(frames[0]);
+const int kFRAME_DISPLAY_TIME = 2;
+
+/*
+ * main object handles Android window frame update, and use webp to decode
+ * pictures
+ */
+class Engine {
+  public:
+    explicit Engine(android_app* app) :
+            app_(app), animating_(false), decoder_(nullptr) {
+        memset(&frameStartTime_, 0, sizeof(frameStartTime_));
+    }
+    ~Engine() {}
+
+    struct android_app* AndroidApp(void) const { return app_; }
+    void StartAnimation(bool start) { animating_ = start; }
+    bool IsAnimating(void) const { return animating_; }
+    void TerminateDisplay(void) { StartAnimation(false); }
+
+     // PrepareDrawing(): Initialize the Engine with current native window geometry
+     //   and blank current screen to avoid garbbage displaying on device
+    bool PrepareDrawing(void);
+
+    // Update webp file into display window when it is decoded
+    // and current frame has been displayed with requested time
+    bool UpdateDisplay(void);
+
+  private:
+    void UpdateFrameBuffer(ANativeWindow_Buffer* buf, uint8_t* src);
+    WebpDecoder* decoder_;
+    struct android_app* app_;
+    bool animating_;
+    struct timespec frameStartTime_;
+};
+
+static int32_t ProcessAndroidInput(struct android_app *app, AInputEvent *event) {
+    Engine* engine = reinterpret_cast<Engine*>(app->userData);
+    if (AInputEvent_getType(event) == AINPUT_EVENT_TYPE_MOTION) {
+        engine->StartAnimation(true);
+        return 1;
+    } else if (AInputEvent_getType(event) == AINPUT_EVENT_TYPE_KEY) {
+        LOGI("Key event: action=%d keyCode=%d metaState=0x%x",
+                AKeyEvent_getAction(event),
+                AKeyEvent_getKeyCode(event),
+                AKeyEvent_getMetaState(event));
+    }
+
+    return 0;
+}
+
+static void ProcessAndroidCmd(struct android_app *app, int32_t cmd) {
+    static int32_t format = WINDOW_FORMAT_RGB_565;
+    Engine* engine = reinterpret_cast<Engine*>(app->userData);
+    switch (cmd) {
+        case APP_CMD_INIT_WINDOW:
+            if (engine->AndroidApp()->window != NULL) {
+                format = ANativeWindow_getFormat(app->window);
+                ANativeWindow_setBuffersGeometry(app->window,
+                              ANativeWindow_getWidth(app->window),
+                              ANativeWindow_getHeight(app->window),
+                              WINDOW_FORMAT_RGB_565);
```

Comment:
compack format, less copying. and it is also more problematic too, discovered that missed a flag to swap endianness for android.  32 is supported, works.